### PR TITLE
add InstallAddon builtin

### PR DESCRIPTION
**Pull request by ronie**
instead of 'forcing' skins to ship with a dozen addons,
give them an option to install those addons on demand.

forum discussion:
http://forum.kodi.tv/showthread.php?tid=231104

### CODE REVIEW

**Comment by MartijnKaijser** [comment]
Question now comes who should throw the install this add-on yes/no dialog, skin or core :)
Blindly doing installation without asking might be a bit too much?

**Comment by phil65** [comment]
Core should throw a yesno dialog askin the user for confirmation in my opinion.
Nice new tool btw, thx for this.

**Comment by ronie** [comment]
it will always show a yes/no popup before installation.
it's enabled by default in core: https://github.com/xbmc/xbmc/blob/master/xbmc/addons/AddonInstaller.cpp#L156

**Comment by HitcherUK** [comment]
Another nice addition, thanks ronie.

File: xbmc/interfaces/builtins/AddonBuiltins.cpp [hunk 1]
```diff
@@ -43,6 +44,30 @@
...
+  {
+    const std::string& addonid = params[0];
+
+    AddonPtr addon;
+    if (!CAddonMgr::GetInstance().GetAddon(addonid, addon))
+    {
+      CAddonInstaller::GetInstance().InstallModal(addonid, addon);
+    }
+    else
+      CLog::Log(LOGERROR, "InstallAddon called with an addonid of an installed addon.");
```
Comment (Razzeee) [inline xbmc/interfaces/builtins/AddonBuiltins.cpp]: I don't think this should be handled as an error, maybe notice would be enough? If we even have that log level?

File: xbmc/interfaces/builtins/AddonBuiltins.cpp [hunk 2]
```diff
@@ -43,6 +44,30 @@
...
+ */
+static int InstallAddon(const std::vector<std::string>& params)
+{
+  if (params.size())
+  {
+    const std::string& addonid = params[0];
+
+    AddonPtr addon;
+    if (!CAddonMgr::GetInstance().GetAddon(addonid, addon))
+    {
```
Comment (Razzeee) [inline xbmc/interfaces/builtins/AddonBuiltins.cpp]: Remove this bracket and the closing one

**Comment by ronie** [comment]
updated as suggested by @Razzeee

**Comment by Razzeee** [comment]
Code seems fine to me.

File: xbmc/interfaces/builtins/AddonBuiltins.cpp [hunk 3]
```diff
@@ -43,6 +44,28 @@
 
 using namespace ADDON;
 
+/*! \brief Install an addon.
+ *  \param params The parameters.
+ *  \details params[0] = add-on id.
+ */
+static int InstallAddon(const std::vector<std::string>& params)
+{
+  if (params.size())
```
Comment (tamland) [inline xbmc/interfaces/builtins/AddonBuiltins.cpp]: Unnecessary. the number you enter in GetOperations ensures the parameter is there

File: xbmc/interfaces/builtins/AddonBuiltins.cpp [hunk 4]
```diff
@@ -43,6 +44,28 @@
...
+  {
+    const std::string& addonid = params[0];
+
+    AddonPtr addon;
+    if (!CAddonMgr::GetInstance().GetAddon(addonid, addon))
+      CAddonInstaller::GetInstance().InstallModal(addonid, addon);
```
Comment (tamland) [inline xbmc/interfaces/builtins/AddonBuiltins.cpp]: Second parameter is an out parameter. so remove line 58

File: xbmc/interfaces/builtins/AddonBuiltins.cpp [hunk 5]
```diff
@@ -43,6 +44,28 @@
...
+    else
+      CLog::Log(LOGNOTICE, "InstallAddon called with an addonid of an installed addon.");
```
Comment (tamland) [inline xbmc/interfaces/builtins/AddonBuiltins.cpp]: This too seems unnecessary to me. Log errors and debug info. There's no need to log that something expected happened..

**Comment by tamland** [comment]
Frankly, Im a bit worried this will just be yet another thing for skins to abuse when you give them tools to explicitly install addon. Wouldn't be better to extend the functionally we already have for plugins that prompts for install when you attempt to use it?

**Comment by jeroenpardon** [comment]
Yet another thing?

Some functions provided by add-ons aren't explicitly started by the user but used by skins to enhance functionality or information. There's no way to prompt for those functions.

This provides a way for the skin to offer "out of the box functionality" without forcing the installation of add-ons. Coupled with a dialog explaining the user why the installation of the add-on is requested provides a much better process than it currently is. A welcome change imo.

**Comment by HitcherUK** [comment]
Plus the fact that these addons can be uninstalled if the user so wishes - something that can't be done if they're required in the addon.xml.

**Comment by ronie** [comment]
thanx for the review @tamland , code updated as suggested.

no worries about the possible abuse. currently we have to 'abuse' runplugin() to install addons, when we don't even want to actually run the addon. we just want to install it.
so let's get a built-in that does exactly that and nothing more.

this can be used for other types of addons as well (resources for instance) that can't be covered by runscript() and the likes.

**Comment by MartijnKaijser** [comment]
jenkins build this please

**Pull request merged**